networkit/cpp/generators/GeometricInhomogenousGenerator.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * GeometricInhomogenousGenerator.h
+ *
+ *  Created on: 09.05.2019
+ *      Author: Manuel Penschuck <networkit@manuel.jetzt>
+ */
+
+#ifndef GEOMETRICINHOMOGENOUS_H_
+#define GEOMETRICINHOMOGENOUS_H_
+
+
+#include <vector>
+#include "StaticGraphGenerator.h"
+
+namespace NetworKit {
+
+class GeometricInhomogenousGenerator: public StaticGraphGenerator {
+public:
+    using coordinate_t = std::vector<double>;
+
+    /**
+     * @param[in] n Number of nodes
+     * @param[in] avgDegree The desired average degree with 0 < avgDegree < n-1
+     * @param[in] exp Target exponent of power-law distribution with exp > 2.0
+     * @param[in] T Temperature with 0 <= T < 1
+     * @param[in] dimension Dimension of the underlying geometry with 1 <= dimension <= 5
+     */
+    GeometricInhomogenousGenerator(count n, double avgDegree, double exp=3, double T=0.0, unsigned dim=1);
+
+    /**
+     * Construct generator from weights that are then scale to match avgDegree and T.
+     *
+     * @param[in] points Coordinates of points
+     * @param[in] weights Unscaled weights (assumed to be powerlaw distributed)
+     * @param[in] avgDegree The desired average degree with 0 < avgDegree < n-1
+     * @param[in] T Temperature with 0 <= T < 1
+     */
+    GeometricInhomogenousGenerator(std::vector<coordinate_t>&& points, std::vector<double>&& weights, double avgDegree, double T);
```

Comment:
Why do the constructors take the vectors by rvalue reference? You're essentially forcing users who want to pass copies to do this explicitly. I would prefer passing the vectors by value. This will still be cheap (at least if the user passes a temporary or uses `std::move`) but will also gracefully fall back to a copy.

Reply:
In order to keep the necessary memory foot print as small as possible, we do the following:
- Move the user provided data set into the container.
- Convert it into an internal data structure
- OPTIONAL: Delete the the user provided data structure.
- Build the graph <- this is when we need the most memory so deleting the input might make sense.

The reason it is moved into the class is to be able delete the data as soon as it's not needed anymore. I you consider this interface as too complicated, we can drop the capability of deleting the data.

Reply:
Yes, I understand that you want to do this. My point is that passing by value enables this behavior without unexpected pitfalls. If you take the vector by value, you still get zero-copy behavior if the users passes a rvalue. However, it also works correctly if tje users passes an lvalue and decides to reuse it later.

Reply:
Thanks. Missed that and will fix it!